FirebaseAdmin/FirebaseAdmin.Tests/Auth/CustomTokenTest.cs

```diff
@@ -0,0 +1,145 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using FirebaseAdmin.Auth.Jwt.Tests;
+using Google.Apis.Auth.OAuth2;
+using Xunit;
+
+namespace FirebaseAdmin.Auth
+{
+    public class CustomTokenTest : IDisposable
+    {
+        public static readonly IEnumerable<object[]> TestConfigs = new List<object[]>()
+        {
+            new object[] { FirebaseAuthTestConfig.DefaultInstance },
+            new object[] { TenantAwareFirebaseAuthTestConfig.DefaultInstance },
+        };
+
+        [Theory]
+        [MemberData(nameof(TestConfigs))]
+        public async Task CreateCustomToken(TestConfig config)
+        {
+            var token = await config.CreateAuth().CreateCustomTokenAsync("user1");
+
+            config.AssertCustomToken(token, "user1");
+        }
+
+        [Theory]
+        [MemberData(nameof(TestConfigs))]
+        public async Task CreateCustomTokenWithClaims(TestConfig config)
+        {
+            var developerClaims = new Dictionary<string, object>()
+            {
+                { "admin", true },
+                { "package", "gold" },
+                { "magicNumber", 42L },
+            };
+
+            var token = await config.CreateAuth().CreateCustomTokenAsync("user2", developerClaims);
+
+            config.AssertCustomToken(token, "user2", developerClaims);
+        }
+
+        [Theory]
+        [MemberData(nameof(TestConfigs))]
+        public async Task CreateCustomTokenCancel(TestConfig config)
+        {
+            var canceller = new CancellationTokenSource();
+            canceller.Cancel();
+            var auth = config.CreateAuth();
+
+            await Assert.ThrowsAsync<OperationCanceledException>(
+                () => auth.CreateCustomTokenAsync("user1", canceller.Token));
+        }
+
+        [Theory]
+        [MemberData(nameof(TestConfigs))]
+        public async Task CreateCustomTokenInvalidCredential(TestConfig config)
+        {
+            var options = new AppOptions
+            {
+                Credential = GoogleCredential.FromAccessToken("test-token"),
+                ProjectId = "project1",
+            };
+            var auth = config.CreateAuth(options);
+
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(
+                () => auth.CreateCustomTokenAsync("user1"));
+
+            var errorMessage = "Failed to determine service account ID. Make sure to initialize the SDK "
```

Comment:
Just a thought, this exact string appears in three other places in the codebase already, maybe if it were a const string the code would be cleaner.

Reply:
I believe it's in `IAMSigner` and a couple of test cases. I'd prefer to have the error message repeated in the tests to make sure developers see the correct error message at runtime (it actually contains some useful information for debugging). But if it starts appearing in more places, we can look into introducing a constant of some sorts.